Cargo clippy changes

elichai opened the pull request:
Hi,
This PR includes a bunch of stuff,
the first thing is to see how people feel about clippy suggestions, unlike rustfmt clippy's suggestions don't change that often and contain a lot of useful suggestions.

Commit e5ad634f3de824253501e812392900b14f8bd13b should increase performance by batching allocations. 
(https://rust-lang.github.io/rust-clippy/master/index.html#slow_vector_initialization, https://rust-lang.github.io/rust-clippy/master/index.html#map_entry)

Commit c9ba8ec9b9b07c89d183533238ccff43903f7f96 might reduce memory usage by removing needless references and pass by value small Copy objects.

Commit ae26c7d0daf73c55f2fcfceb534f4e26ef50fb9e is one that I think is important, `PartialEq` and `Hash` *must* agree (meaning that `a==a` means that `Hash(a)==Hash(a)`) so it is really discouraged to manually implement one but derive the other, 
and since we didn't do anything special there I preferred to derive them both. 
(https://rust-lang.github.io/rust-clippy/master/index.html#derive_hash_xor_eq)

Commit bf82f63d3b911b883cf2400d20f1a01350afa11b contains idiomatic suggestions, some *might* improve performance (like using char instead of strings in `split()`), and some *might* improve reliability (using `b'a'` instead of `'a' as u8`) 
(https://rust-lang.github.io/rust-clippy/master/index.html#single_char_pattern https://rust-lang.github.io/rust-clippy/master/index.html#char_lit_as_u8 )

~~The last commit 6e234f7d2031b443e666cbca63b476b112bcd840 I'm not sure how I feel about it I decided to add it, though it can be remove if wanted.~~ (dropped)

dpc commented on src/blockdata/script.rs:
```diff
@@ -446,11 +446,9 @@ impl<'a> Iterator for Instructions<'a> {
-                        self.data = &[];
-                        return Some(Instruction::Error(Error::NonMinimalPush));
-                    }
+                if self.enforce_minimal && n == 1 && (self.data[1] == 0x81 || (self.data[1] > 0 && self.data[1] <= 16)) {
```
I think in this case `clippy` is wrong, and I don't like this lint. Having a top `if self.enforce_minimal {` makes it clear, that the whole block of code applies only if `self.enforce_minimal`, without having to scan the whole expression.

I'd recommend applying an override instead to mute `clippy` here.

elichai replied:
This PR doesn't make the repo clippy compliant yet. we'll of course need to mute a bunch of suggestions, this is just a first try to see how people feel about this and to at least integrate the important suggestions.

I do think it's still readable because it short circuits, maybe having a new line there could help readability.

apoelstra replied:
I also think the original was more readable.

elichai replied:
reverted

dpc commented on src/util/amount.rs:
```diff
@@ -135,14 +135,14 @@ fn parse_signed_to_satoshi(
     }
 
-    let negative = s.chars().next().unwrap() == '-';
+    let negative = s.starts_with('-');
```
It's not problem with this patch, but since I'm looking at it: I'd name this variable `is_negative` so it's clear its a boolean.

elichai replied:
Done in #309

dpc commented on src/util/bip32.rs:
```diff
@@ -180,13 +180,10 @@ impl FromStr for ChildNumber {
-            false => ChildNumber::from_normal_idx(
-                inp.parse().map_err(|_| Error::InvalidChildNumberFormat)?
-            )?,
+        Ok(if inp.chars().last().map_or(false, |l| l == '\'' || l == 'h') {
```
Seems to me this should be extracted into a small function to make it clear what's the purpose here (`is_hardened_path (inp)` or smth).

stevenroose replied:
Or just a variable. Variable names can add a lot of clarity.
```rust
let is_hardened = ...;
Ok(if is_hardened { ...
```

elichai replied:
Done

apoelstra commented on src/util/amount.rs:
```diff
@@ -608,9 +608,10 @@ impl SignedAmount {
+        Ok(if negative {
+            SignedAmount(-(satoshi as i64))
+        } else {
+            SignedAmount(satoshi as i64)
```
I don't feel strongly about this, but why is 5 lines better than 4?

elichai replied:
supposedly matching over a bool is less readable. I think that sometimes match can be more readable, don't have strong feelings about this too https://rust-lang.github.io/rust-clippy/master/index.html#match_bool

apoelstra commented on src/util/bip32.rs:
```diff
@@ -180,13 +180,10 @@ impl FromStr for ChildNumber {
-                inp.parse().map_err(|_| Error::InvalidChildNumberFormat)?
-            )?,
+        Ok(if inp.chars().last().map_or(false, |l| l == '\'' || l == 'h') {
+            ChildNumber::from_hardened_idx(inp[0..inp.len() - 1].parse().map_err(|_| Error::InvalidChildNumberFormat)?)?
```
What is our line length limit?

This also feels harder to read than the original.

elichai replied:
I felt like both this and the original are unreadable, but I didn't want to actually format code in this PR, I think that after formatting the if will be more readable

stevenroose replied:
Default Rust line length limit is 100 chars. That's also what I used in https://github.com/rust-bitcoin/rust-bitcoin/pull/290.

apoelstra replied:
How would people feel about dropping it to 80?

elichai replied:
Personally I don't like short lines. especially in function declarations/call. I hated this so much in PEB8 when I wrote python.
I think we should use common sense. 
when you do stuff like the builder pattern than adding a space for every function is good, when you declare a function than passing the max length by 5-10 chars is fine in my opinion, we all have FHD/QHD+ displays here.

Anyway this is a rustfmt thing, not a clippy thing.

apoelstra replied:
> we all have FHD/QHD+ displays here

But some of us like to view multiple files at once ;)

Anyway sure, I'm fine with 100 lines. I'm just annoyed when rustfmt does stuff like collapsing a "common sense" multi-line builder pattern into a 100-line mess.

elichai replied:
yeah, even in my personal projects where I have a `rustfmt.toml` file I still find myself manually overriding stuff that it do

dpc replied:
@apoelstra It's possible to tweak this, I belive. `use_small_heuristics=off` https://rust-lang.github.io/rustfmt/ ?

stevenroose replied:
I use `use_small_heuristic = "Off"` in all my projects. I don't like rustfmt to "heuristically" decide whether a line "feels" short/long.

I'm personally not in favor of shortening it to 80, though. Sadly rustfmt doesn't support a separate line length for comments, where I do prefer 80.

80 is just really too short for any method declaration involving `Result`..

apoelstra commented:
NAK the "remove redundant field names". This is one of the worst misfeatures in Rust's syntax.

dpc commented:
> NAK the "remove redundant field names". This is one of the worst misfeatures in Rust's syntax.

@apoelstra Why? All this redundancy serves no purpose, and I can't see how it helps anything or prevents making a mistake.

apoelstra commented:
It's not redundancy. A field name and a variable are completely different kinds of things, even at the parse level.

Conflating them would make things harder to read/understand, even if it was done without an adhoc separate parse rule that doesn't exist in any other language.

elichai commented:
Yeah I have mixed feelings about that feature.
I think it's only useful for readability when *all* the fields are the same as the variable name, I don't like it when it's mixed.

elichai commented:
Dropped the last commit. (implicit field names)

dpc commented:
> It's not redundancy. A field name and a variable are completely different kinds of things, even at the parse level.

What do I/you care? It's a problem of a compiler (or Rust compiler developers). :rofl: 

The way I think about it:  Awfully a lot of times the name of the field is going to be the same as the name of variable used to initialize it. Because they are holding the same thing, right? And they are of the same type, size, performance, flavor and smell.

In essence in:

```
let foo = make_foo();
struct FooBar
    foo: foo,
}
```

the `foo` variable *becomes* `foo` field. So why type it twice? Why make humans look at it twice?

The cost: well, there's a new special case rule that might seem odd at first. But after you accept it into your life, it's not actually a problem. The benefit: code looks more compact, less typing, less room for typos.

This case is so common, then I think it's a worthwhile tradeoff. And after a while I kind of think that `foo: someexpression` is actually a special case.

apoelstra commented:
I care because I need to parse the code to read it.

You're assigning the local variable `foo` to the field with name `foo`. These two uses of `foo` are not remotely in the same category, and neither is in the same category as `foo: foo` which is an assignment.

It's extremely jarring to see something that looks like one of the former, when I'm expecting a field name, then upon seeing a `,` rather than `:` I need to mentally compute and substitute the latter.

dpc commented:
>  I need to mentally compute and substitute the latter.

Give it some time. :)

apoelstra commented:
Fine :). I concede.

stevenroose commented:
Except for those few occasions where a long squashed line is produced, I like all the changes.

elichai commented:
After #349 is merged I'll rebase this and leave only the non controversial ones, (some of these things have theoretic performance effects).

elichai commented:
Rebased. 
removed all of the controversial changes, the only change that might be controversial IMHO is the use of `std::cmp::Ordering`, how do people feel about it?

elichai commented:
Oh, another thing that I'm not 100% sure about is the use of the entry API,
because we need the key only if the key is occupied we can't get it by value (looked into changing this is libstd, but that's not possible because of `Entry::insert` which needs the key by value)

anyway this is a tradeoff between re-looking into the map(BTree is log n for each operation) and cloning the heap allocated key.

stevenroose commented:
> anyway this is a tradeoff between re-looking into the map(BTree is log n for each operation) and cloning the heap allocated key.

I think in most cases, cloning is preferred. Keys are usually short, `Copy` and stack allocated. Like hashes.

stevenroose commented:
ACK 4793975036b3745cbcb4628440e1f9d6a22a8984

apoelstra commented:
Needs rebase

stevenroose commented on src/util/amount.rs:
```diff
@@ -131,7 +132,7 @@ impl error::Error for ParseAmountError {
 
 fn is_too_precise(s: &str, precision: usize) -> bool {
-    s.contains(".") || precision >= s.len() || s.chars().rev().take(precision).any(|d| d != '0')
+    s.contains('.') || precision >= s.len() || s.chars().rev().take(precision).any(|d| d != '0')
```
Does this change the `Pattern` from a `&str` to a `char`? Is that more efficient? Neat, though.

elichai replied:
asm: https://godbolt.org/z/bLK8Hx

The exact code of `Pattern` is complicated by the intuition is as follows:
if char is less than a byte(ie ascii) so you can use `slice::contains` which basically iterates over the slice until it finds it. (`memchr`)
if it's higher then it defaults to the `str` impl which creates a haystack and does a bunch of logic to calculate UTF-8 bytes points

stevenroose commented on src/util/psbt/map/global.rs:
```diff
@@ -13,6 +13,7 @@
 //
 
 use std::collections::BTreeMap;
+use std::collections::btree_map::Entry;
```
Why do you import the Entry type here but not in the previous file?

stevenroose replied:
Nor in the next it seems.

elichai replied:
When it was part of a macro I used the full path as we usually do in macros, when it wasn't part of a macro I imported it for better readability.

stevenroose replied:
ah ok, that makes perfect sense
